HarithJ/ci-cd-2019
# Story/ct 1 create pod

#### Description
This PR would add in the following packages:
1. **github/webhook**: This package would be responsible for all the functionalities associated with Github's webhook. In this PR, it adds in a `Payloadfunc` handler that would be hit when Github's webhook is triggered. This handler is responsible to read Github's payload and reduce it to the fields that would be useful to us for cloning the repository.

2. **github/payload**:  This package contains two structs for the Github's payload. One struct is where the payload would get unmarshalled and the other would simplify the payload.

3. **github/git**: This package is responsible for all the functionalities associated with git. In this PR, it contains a function to clone the repository.

4. **kubernetes/auth**: This package has the functionalities to authenticate clusters. In this PR, it authenticates into the cluster it is running in.

5. **kubernetes/pod**: This package would hold all the functionalities to do with k8s pods. At this stage, it has the functionality to create a pod.

6. **pipeline/pipeline**: This has all the functionalities to do with the pipeline. It contains a function to start a pipeline by first creating a pod.

This PR would also turn the application to an API and serves two endpoints at the moment:
1. **Index endpoint (`/`)**: Displays the welcome text

2. **Payload endpoint (`/payload`)**: Responsible for Github's webhooks.

#### Type of change


#### How Has This Been Tested?

This has been tested locally via minikube, if you do not have minikube installed, you can use GCP or any other kubernetes platform:
1. Apply both the files located in kubernetes folder.

2. Make sure you have the k8s deployment URL exposed, if you are running on minikube, you can expose the deployment via by running the following command:
`kubectl expose deployment ci-cd-2019 --type=NodePort`
and then use `ngrok` to make the API accessible over the internet.
 
3. Copy the URL or IP address of your application and insert it as one of the Github webhooks ending with `/payload`. Make sure to change the `content type` to `application/json`. Allow Pull Requests to trigger this webhook only.

4. Checkout to another branch, create a folder named 2019 and a file named pod.yml in it with pod configurations you would like to create. Finally, raise a PR.

NOTE: you can use this https://github.com/HarithJ/test-ci-cd-2019 repo to test it out.

#### Checklist:
N/A

#### JIRA
https://andela-apprenticeship.atlassian.net/browse/CT-1
